### PR TITLE
keep dependency path in static linkage check

### PR DESCRIPTION
**Pull request by elharo**
@suztomo @netdpb This shows us the paths by which we reached a jar that has static linkage errors.

### CODE REVIEW

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java [hunk 1]
```diff
@@ -91,12 +91,15 @@ public static Path generate()
     ArtifactCache cache = loadArtifactInfo(managedDependencies);
     
     ImmutableList<Path> classpath = StaticLinkageChecker.artifactsToClasspath(managedDependencies);
```
Comment (netdpb) [inline dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: Do you still use classpath?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java [hunk 2]
```diff
@@ -67,6 +67,7 @@
 
   static ClassDumper create(List<Path> jarFilePaths) throws IOException, ClassNotFoundException {
     // Creates classpath in the same order as inputClasspath for BCEL API
+
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java]: Why this change?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java [hunk 3]
```diff
@@ -28,19 +29,25 @@
   /**
    * Returns the absolute path of the jar file containing source classes of linkage errors
    */
-  public abstract Path getJarPath();
+  abstract Path getJarPath();
+  
+  /**
+   * @return the paths to this artifact
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java]:
```suggestion
   * Returns the paths to this artifact.
```

https://google.github.io/styleguide/javaguide.html#s7.2-summary-fragment

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java (hunk 2 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java]: random junk as I edited

File: dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java (hunk 1 above)
Comment (elharo) [reply dashboard/src/main/java/com/google/cloud/tools/opensource/dashboard/DashboardMain.java]: yes, on line 97. I should get rid of this but my first attempt broke things.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java (hunk 3 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java]: done

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java [hunk 4]
```diff
@@ -53,6 +57,28 @@ public String toString() {
...
+    
+    Artifact grpcArtifact = new DefaultArtifact("io.grpc:grpc-auth:1.15.1");
+    ListMultimap<Path, DependencyPath> multimap =
+        StaticLinkageChecker.artifactsToPaths(ImmutableList.of(grpcArtifact));
+
+    Set<Path> paths = multimap.keySet();
+    
+    Truth.assertThat(paths)
+        .comparingElementsUsing(PATH_FILE_NAMES)
+        .contains("grpc-auth-1.15.1.jar");
```
Comment (netdpb) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java]:
```suggestion
        .containsAllOf("grpc-auth-1.15.1.jar", "google-auth-library-credentials-0.9.0.jar");
```

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java [hunk 5]
```diff
@@ -122,7 +148,7 @@ public void testFindInvalidReferences_selfReferenceFromAbstractClassToInterface(
     SymbolReferenceSet symbolReferenceSet = ClassDumper.scanSymbolReferencesInJar(httpClientJar);
 
     JarLinkageReport jarLinkageReport =
-        staticLinkageChecker.generateLinkageReport(httpClientJar, symbolReferenceSet);
+        staticLinkageChecker.generateLinkageReport(httpClientJar, symbolReferenceSet, null);
```
Comment (netdpb) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java]: I don't think `generateLinkageReport()` should accept null. Instead pass an empty list or something.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 6]
```diff
@@ -51,14 +56,21 @@
...
-        reportOnlyReachable, ClassDumper.create(jarFilePaths), entryPoints);
+    
+    ClassDumper dumper = ClassDumper.create(jarFilePaths);
+    return new StaticLinkageChecker(onlyReachable, dumper, entryPoints);
+  }
+  
+  public static StaticLinkageChecker create(boolean onlyReachable,
+      ListMultimap<Path, DependencyPath> paths, ImmutableSet<Path> entryPoints) 
+          throws ClassNotFoundException, IOException {
+    List<Path> jarFilePaths = new ArrayList<Path>();
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: done

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 7]
```diff
@@ -51,14 +56,21 @@
...
-        "The linkage classpath is empty. Specify input to supply one or more jar files");
-    return new StaticLinkageChecker(
-        reportOnlyReachable, ClassDumper.create(jarFilePaths), entryPoints);
+    
+    ClassDumper dumper = ClassDumper.create(jarFilePaths);
+    return new StaticLinkageChecker(onlyReachable, dumper, entryPoints);
+  }
+  
+  public static StaticLinkageChecker create(boolean onlyReachable,
+      ListMultimap<Path, DependencyPath> paths, ImmutableSet<Path> entryPoints)
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: Key order does matter. It's already a LinkedListMultimap further down. Made that epxlicit.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 8]
```diff
@@ -51,14 +56,21 @@
...
+    ClassDumper dumper = ClassDumper.create(jarFilePaths);
+    return new StaticLinkageChecker(onlyReachable, dumper, entryPoints);
+  }
+  
+  public static StaticLinkageChecker create(boolean onlyReachable,
+      ListMultimap<Path, DependencyPath> paths, ImmutableSet<Path> entryPoints) 
+          throws ClassNotFoundException, IOException {
+    List<Path> jarFilePaths = new ArrayList<Path>();
+    jarFilePaths.addAll(paths.keySet());
+    ClassDumper dumper = ClassDumper.create(jarFilePaths);
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: seems ClassDumper really does need a List, not an Iterable, in order to stream it.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 9]
```diff
@@ -70,12 +82,23 @@ public static StaticLinkageChecker create(
   private final ClassDumper classDumper;
 
   private final ImmutableSet<Path> entryPoints;
+  
+  private final ListMultimap<Path, DependencyPath> paths;
 
   StaticLinkageChecker(
       boolean reportOnlyReachable, ClassDumper classDumper, Iterable<Path> entryPoints) {
     this.reportOnlyReachable = reportOnlyReachable;
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: done

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 10]
```diff
@@ -70,12 +82,23 @@ public static StaticLinkageChecker create(
...
     this.classDumper = classDumper;
     this.entryPoints = ImmutableSet.copyOf(entryPoints);
+    this.paths = ArrayListMultimap.create();
+  }
+  
+  StaticLinkageChecker(boolean reportOnlyReachable, ClassDumper classDumper,
+      Iterable<Path> entryPoints, ListMultimap<Path, DependencyPath> paths) {
+    this.reportOnlyReachable = reportOnlyReachable;
+    this.classDumper = classDumper;
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: done

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 11]
```diff
@@ -112,30 +135,46 @@ public static void main(String[] arguments)
    */
   public static ImmutableList<Path> artifactsToClasspath(List<Artifact> artifacts)
       throws RepositoryException {
+    
+    ListMultimap<Path, DependencyPath> multimap = artifactsToPaths(artifacts);
+    LinkedHashSet<Path> set = new LinkedHashSet<>(); // to remove duplicate keys
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: done

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 12]
```diff
@@ -112,30 +135,46 @@ public static void main(String[] arguments)
...
+    for (Entry<Path, DependencyPath> entry : multimap.entries()) {
+      set.add(entry.getKey());
+    }
+    
+    return ImmutableList.<Path>builder().addAll(set).build();
+  }
+  
+  
+  // Multimap is a pain, maybe just use LinkedHashMap<Path, List<DependencyPath>>
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: Because it's ~8 different types with unclear behavior around whether key and/or value order is or is not preserved that I can never remember without checking the Javadocs and can't get right without really solid test coverage. At the end of the day, it's just too complex.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 13]
```diff
@@ -112,30 +135,46 @@ public static void main(String[] arguments)
...
+   * Finds jar file paths for Maven artifacts and their dependencies.
+   *
+   * @param artifacts Maven artifacts to check
+   * @return map absolute paths of jar files to Maven dependency paths
+   * @throws RepositoryException when there is a problem in retrieving jar files
+   */
+  public static ListMultimap<Path, DependencyPath> artifactsToPaths(List<Artifact> artifacts)
+      throws RepositoryException {
+    
+    ListMultimap<Path, DependencyPath> multimap = LinkedListMultimap.create();
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: Made it an explicit LinkedListMultimap instead.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 14]
```diff
@@ -174,8 +215,16 @@ public StaticLinkageCheckReport findLinkageErrors() throws ClassNotFoundExceptio
    * @return linkage report for the jar file, which includes linkage errors if any
    */
   @VisibleForTesting
-  JarLinkageReport generateLinkageReport(Path jarPath, SymbolReferenceSet symbolReferenceSet) {
-    JarLinkageReport.Builder reportBuilder = JarLinkageReport.builder().setJarPath(jarPath);
+  JarLinkageReport generateLinkageReport(Path jarPath, SymbolReferenceSet symbolReferenceSet,
+      Iterable<DependencyPath> dependencyPaths) {
+
+    if (dependencyPaths == null) {
```
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: done

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java (hunk 4 above)
Comment (elharo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java]: done

File: dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java (hunk 5 above)
Comment (elharo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/classpath/StaticLinkageCheckerTest.java]: done

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java (hunk 2 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ClassDumper.java]: Reverting a file is a much more onerous operation in git than perforce. I also wish we could configure Github like we do critique to not even show diffs in trailing whitespace.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (netdpb) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: Then it could take `Collection`, or even `ImmutableSet`.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (netdpb) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: The `Immutable*` types always preserve insertion order. But I guess it's up to you.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 15]
```diff
@@ -65,9 +67,9 @@ static StaticLinkageChecker create(
   }
   
   public static StaticLinkageChecker create(boolean onlyReachable,
-      ListMultimap<Path, DependencyPath> paths, ImmutableSet<Path> entryPoints) 
+      LinkedListMultimap<Path, DependencyPath> paths, ImmutableSet<Path> entryPoints) 
           throws ClassNotFoundException, IOException {
-    List<Path> jarFilePaths = new ArrayList<Path>();
+    List<Path> jarFilePaths = new ArrayList<>();
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: `new ArrayList<>(paths.keySet());`

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java [hunk 16]
```diff
@@ -87,18 +89,15 @@ public static StaticLinkageChecker create(boolean onlyReachable,
...
   }
   
   StaticLinkageChecker(boolean reportOnlyReachable, ClassDumper classDumper,
       Iterable<Path> entryPoints, ListMultimap<Path, DependencyPath> paths) {
     this.reportOnlyReachable = reportOnlyReachable;
-    this.classDumper = classDumper;
-    this.entryPoints = ImmutableSet.copyOf(entryPoints);
-    this.paths = paths;
+    this.classDumper = Preconditions.checkNotNull(classDumper);
+    this.entryPoints = ImmutableSet.copyOf(Preconditions.checkNotNull(entryPoints));
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: You don't need this one. `ImmutableSet.copyOf(null)` will throw.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java (hunk 15 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: done

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: it looks like order matters here. List seems like the right type in this case.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java (hunk 16 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: done

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/StaticLinkageChecker.java]: That's a surprise. To me "Immutable" means you can't add or remove from the collection. It doesn't suggest to me that an unordered type like set has a predictable iteration order, much less that it preserves insertion order. It feels more natural to specify a type that is ordered by definition, i.e. a list, when we want to preserve order.
